### DISCUSSION TITLE
XPRA PAM encryption

### DISCUSSION THREAD

**Discussion by ianblum4** [Xpra-org/xpra#4273] [General]
Hello, I have tried this command `xpra start --start=xterm -d auth --bind-tcp=0.0.0.0:14505,auth=pam`. When using an SSH tunnel PAM works correctly, however but when the username is different and the password is correct, the PAM ignores it and does not check the username. In other words, a user could input a fake username to the XPRA client and the correct password and it incorrectly authenticates them. Futhermore, when using the XPRA client with the command run on the Ubuntu Server, it gets an "ENCRYPTION" error, and I am unable to run PAM with AES Encryption.

I tried this command
`xpra start :100 --bind-tcp=0.0.0.0:14511 --tcp-auth=pam --tcp-encryption=AES --tcp-encryption-keyfile=aes_keyfile`
Also the command
`xpra start :100 --bind-tcp=0.0.0.0:14511,auth=pam --tcp-encryption=AES --tcp-encryption-keyfile=aes_keyfile`
![encryption_error](https://github.com/Xpra-org/xpra/assets/170835285/f84c8387-f5a3-4794-b3a0-08cc14682cb4)

ERROR in LOG:
```
2024-07-01 22:28:38,986 sending challenge "password for user 'joe"
2024-07-01 22:28:39,006 client has requested disconnection: invalid digest
```

**Comment by totaam** [comment]
There seems to be a lot of confusion.
PAM is used for authentication, not encryption.

If you're using an SSH tunnel, then you're likely not using the TCP socket at all.

**Reply by ianblum4** [reply]
Correct. As a clarification, when I try to access the port over the net on the XPRA client, I get an "ENCRYPTION" error. However, when I use the HTML client such as going to the server address on Google Chrome, it works properly, but on the XPRA Windows Client, I get "ENCRYPTION" as the error. How do I fix this?

**Reply by ianblum4** [reply]
I understand that PAM is for authentication, but I am getting an ENCRYPTION error when trying to access over the net using the XPRA windows application. When I open Google Chrome over the net it works fine and authenticates. I figured that I needed use both Encryption along side with PAM authentication to remove this bug. Do you know how to fix the ENCRYPTION error? Hope this helps clarify things. Thank you.

**Reply by totaam** [reply]
xpra correctly refuses to use `pam` authentication without some form of encryption.
Your server log should be showing:
```
Error: authentication failed:
 server requested 'xor' digest, cowardly refusing to use it without encryption
```
This is not a bug.

There is a bug however in that the html5 client should not be letting you use password authentication over insecure connections without the first enabling the "insecure connections" checkbox.

Please see the documentation for all encryption connection options.

**Reply by ianblum4** [reply]
Ok, I tried to use an AES encryption, but then it produces another bug which doesn't take PAM authentication into account and allows any user to connect, even if they do not have an account. I tried looking at documentation for hours. If you know the proper command I should use to use PAM and AES encryption I would greatly appreciate this. Thank you.
xpra start :100 --bind-tcp=0.0.0.0:14511,auth=pam --tcp-encryption=AES --tcp-encryption-keyfile=aes_keyfile
T

**Reply by ianblum4** [reply]
To further clarifiy, I want to know the specific command to use AES encryption with PAM, I tried this xpra start :100 --bind-tcp=0.0.0.0:14511 --tcp-auth=pam --tcp-encryption=AES --tcp-encryption-keyfile=aes_keyfile.
But it ignored PAM, and allowed any user to connect with fake username and fake password, massive security risk. It seems that PAM was only being used when stopping the server. Thank you again for assisting in this. Spent numerous hours searching the internet, this forum, and the documentation.

**Reply by ianblum4** [reply]
When using the command I attached with the auth=pam instead of --tcp-auth=pam, I get
2024-07-01 23:52:24,611 Warning: authentication failed
2024-07-01 23:52:24,611  missing encryption tokens from client
I ran using the XPRA client on Windows
I also tried using the attach command and told it where the aeskey was located. It still did not work. Not sure what I am missing

**Reply by totaam** [reply]
> but then it produces another bug which doesn't take PAM authentication into account

I very much doubt that.

All the authentication and encryption can be found here:
https://github.com/Xpra-org/xpra/tree/master/docs/Network
And you're not using the same syntax, so likely not following these instructions.

You're also not specifying which version you have, OS, etc..

**Reply by ianblum4** [reply]
Unfortunately, it is producing a bug. " xpra start --start=code -d auth --bind-tcp=0.0.0.0:14513 --encryption=AES --encryption-keyfile=/home/ianblum1/aes_keyfile --auth=pam" allows any user to gain access to my machine. A big security hole. 
2024-07-02 00:00:39,094 Warning: client expects an authentication challenge,
2024-07-02 00:00:39,094  sending a fake one 2024-07-02 00:00:40,052 Python/GTK3 Microsoft Windows aero client version 5.0
2024-07-02 00:00:40,052  connected as 'icebe' my username is ianblum1 not icebe. I am running XPRA 5.0.9 on a  Ubuntu 20.04.6 LTS Server. My Windows Client is Windows 11. I can confirm my AES key is valid. It is quite possible my syntax of this command is incorrect which is causing the issue. The documentation does not show how to use both an encryption and PAM together. This command for AES does not include for PAM:
xpra start --start=xterm \
     --bind-tcp=0.0.0.0:10000,encryption=AES,keyfile=key.txt
client:
xpra attach "tcp://localhost:10000/?encryption=AES&keyfile=./key.txt"


I want to use the interactive XPRA client on my windows machine and be able to sign in using PAM and an encryption model thats supported. I tried ssl as well and could not get it to work. Do you know the specific command I need? I really appreciate the assistance. To further explain I have spent at least 8 hours trying to resolve this issue. If you know the specific command I would greatly appreciate it.

**Reply by ianblum4** [reply]
I'm happy to produce more information. I appreciate your time.

**Reply by totaam** [reply]
You're using the wrong syntax, again.
Please use the syntax from the documentation.

**Reply by ianblum4** [reply]
I just copied and passed the syntax from the documentation. 
```shell
xpra start --start=xterm --bind-tcp=0.0.0.0:10000,encryption=AES,keyfile=key.txt
```
client:
```shell
xpra attach "tcp://localhost:10000/?encryption=AES&keyfile=./key.txt"
```
This is directly from your website.
![image](https://github.com/Xpra-org/xpra/assets/170835285/7ef65d35-0c3d-4c5f-9e81-cf8fba0b4639)
Here is for PAM:
```shell
xpra start --start=xterm -d auth --bind-tcp=0.0.0.0:10000,auth=pam
```
![image](https://github.com/Xpra-org/xpra/assets/170835285/9d510053-d47e-4419-82a6-bc748caf2de6)

Please inform me how to tie them both together so I have a working XPRA program. Thanks.

**Reply by totaam** [reply]
Sorry, but that's just too many messages to read, not enough responses and some fairly wild claims of "bugs" everywhere.
So I'm taking a break, I will get back to this later.

**Reply by ianblum4** [reply]
Sure, thanks for helping.

**Reply by ianblum4** [reply]
Hey Antoine hope you had a nice holiday and weekend. Any updates on this? Your help is greatly appreciated

**Comment by totaam** [comment]
First things first:
> but when the username is different and the password is correct, the PAM ignores it and does not check the username

That's correct and working as intended.
With the default settings, some authentication modules do not allow remote users to try any username they like and restrict authentication to the currently logged in username only:
https://github.com/Xpra-org/xpra/blob/39527ec7fd0c6b07650d150fe52a738e293b6e21/xpra/server/auth/sys_auth_base.py#L77-L80
This is the case for `pam` (aka `sys`).
To override this behaviour, use `auth=pam,client-username=true`.
But that's generally not what you want unless you are running a proxy server.

> This is directly from your website.

This is not my website, it is the project's documentation.

Please do not paste pictures of text, paste text so that it remains searchable.

> I just copied and passed the syntax from the documentation.

The AES example works for me.

Adding `auth=sys` (or `pam`) works just as well, except I have to provide my system password to authenticate with the session.

Please note that the syntax in the documentation has been updated for v6.x: 39527ec7fd0c6b07650d150fe52a738e293b6e21
(this should not make any difference to `pam` or `sys` since you were not using extra attributes in your command lines)

---

If you are still having problems, please read https://github.com/Xpra-org/xpra/wiki/Reporting-Bugs before replying.

**Reply by totaam** [reply]
See also https://github.com/Xpra-org/xpra/issues/4294

**Reply by ianblum4** [reply]
The AES example is giving me issues.  Where should I have the aes_key stored on my windows machine to ensure that it works properly? I got an invalid "AES" error with the attach command.

**Reply by totaam** [reply]
You haven't read the link, have you?
No command line pasted, no exact error message, no OS or version information, ...

**Reply by ianblum4** [reply]
Hi I have read the link. Sorry for not including the specific information. Thank you for your help.
I am using a windows machine to use the client and I used the AES example and it is working now. 
However I want to use PAM with AES and I am getting the following error, I am running XPRA on a ubuntu machine and get the same error on versions 5.0.9 and 6.1 beta 
```
2024-07-10 12:43:07,381 sending data using AES-CBC encryption
2024-07-10 12:43:07,384 receiving data using AES-CBC encryption
2024-07-10 12:43:07,384 Authentication required by PAM authenticator module
2024-07-10 12:43:07,384  sending challenge using 'xor' digest over tcp connection
2024-07-10 12:43:07,392 Disconnecting client ip address:portnumber
2024-07-10 12:43:07,392  unencrypted packet dropped
```
This is the command I am using `xpra start --start=code --bind-tcp=0.0.0.0:portnumber,encryption=AES,keyfile=keyfile.txt,auth=pam,client-username=true` on the Ubuntu Server.
This is the command I am using on my windows machine to launch the tool: 
`xpra attach tcp://username:password@severipaddress:portnumber/?encryption=AES&keyfile="aes_keyfile.txt"`

I can confirm this is working:
`xpra start --start=xterm --bind-tcp=0.0.0.0:10000,encryption=AES,keyfile=key.txt`

Thank you again for your help.

**Reply by totaam** [reply]
On MS Windows, use "Xpra_cmd.exe" to see the command output easily, otherwise you have to open the log file afterwards to see the messages.
My guess is that `aes_keyfile.txt` is not in the directory where xpra is installed (EXE installer? ZIP? version?), so you would need to use a full path instead.
Have you tried from a Linux machine, for example the server itself?
`xpra info 'tcp://username:password@127.0.0.1:portnumber/?encryption=AES&keyfile=aes_keyfile.txt'`

**Reply by ianblum4** [reply]
Hey, just to reiterate the AES encryption example is working correctly but when used with PAM it is not working. Used this command on my Linux Machine: `xpra start --start=code --bind-tcp=0.0.0.0:port,encryption=AES,keyfile=/home/ianblum1/aes_keyfile.txt,auth=pam,client-username=true`
```
C:\Program Files\Xpra>xpra_cmd.exe attach "tcp://user:password@ip:portnumber/?encryption=AES&keyfile="aes_keyfile.txt"

2024-07-11 23:07:26,645 Xpra GTK3 client version 6.0.1-r0
2024-07-11 23:07:27,355  running on Microsoft Windows Microsoft Windows 11 Home
2024-07-11 23:07:27,363  cpython 3.11
2024-07-11 23:07:32,329 GStreamer version 1.24.3
2024-07-11 23:07:32,355 created named pipe 'Xpra\41652'
2024-07-11 23:07:32,800 keyboard layout 'United States - English' : 'us' (0x409)
2024-07-11 23:07:34,071 OpenGL accelerate missing: numpy_formathandler
2024-07-11 23:07:34,100 OpenGL enabled on 'Intel(R) Iris(R) Xe Graphics'
2024-07-11 23:07:34,482 receiving data using AES encryption
2024-07-11 23:07:34,545 Error: authentication failed:
2024-07-11 23:07:34,545  server requested 'xor' digest, cowardly refusing to use it without encryption
```
Is my syntax correct for PAM?

**Comment by totaam** [comment]
With git master 6.2 beta as of b16a48f6137e08f43fc38b4d96d16d49b3392ed7:
```shell
xpra start --start=xterm --bind-tcp=0.0.0.0:10000,encryption=AES,keyfile=key.txt,auth=allow --no-daemon
```
```
xpra info 'tcp://foo:bar@127.0.0.1:10000/?encryption=AES&keyfile=key.txt'
```
Works and the server shows: `sending data using AES-CBC encryption` and `receiving data using AES-CBC encryption`

Changing `auth=allow` to `auth=pam` - but without using the correct password with `xpra info`, as expected:
```
Warning: pam authentication failed: AUTH_ERR
 Authentication failure
```
Using the correct password for the user running the `xpra start` command:
```
pam_authenticate: SUCCESS
```
And the command succeeds, showing that the username `foo` is being ignored.

Adding `client-username=true`, the authentication fails again as `foo` is not a valid username.
Specifying the correct username, succeeds.

Running the seamless server as root (dangerous!) and adding `verify-username=false`, I can authenticate using any valid username + password combination.

Note: b16a48f6137e08f43fc38b4d96d16d49b3392ed7 is probably the correct fix for `xor`.
